### PR TITLE
Terry/cal flash pattern

### PR DESCRIPTION
**Pull request by kplus**

### CODE REVIEW

File: host/bloodview/src/data-cal.c [hunk 1]
```diff
@@ -198,45 +253,55 @@ static uint32_t data_cal__data_channel_to_acq_channel(
...
 		}
 
-		data_cal__calibrate_channel(
+		src = c->src;
+		if (src >= BL_ACQ_SOURCE_MAX) {
+			continue;
+		}
+
+		uint32_t hw_scale =	main_menu_config_get_source_hw_oversample(src) -
+							main_menu_config_get_source_hw_shift(src);
```
Comment (tlsa) [inline host/bloodview/src/data-cal.c:284]: Indentation looks weird.

```suggestion
		uint32_t hw_scale = main_menu_config_get_source_hw_oversample(src) -
				main_menu_config_get_source_hw_shift(src);
```

File: host/bloodview/src/data.c [hunk 2]
```diff
@@ -192,13 +192,13 @@ static bool data__handle_sample(
 	unsigned index = channel->sample_count & DATA_MASKS_MASK;
 
 	if (data_g.sample_masks[index] & (1u << acq_channel)) {
-		fprintf(stderr, "Data error: Channel %u sample overrnun\n",
+		fprintf(stderr, "Data error: Channel %u sample overrun\n",
```
Comment (tlsa) [inline host/bloodview/src/data.c:195]: Ha!  Nice :)

File: host/bloodview/src/device.c [hunk 3]
```diff
@@ -557,13 +557,8 @@ static bool device__queue_msg_led(bool enable_lights)
...
 
-/**
- * Get the appropriate source for a given channel, according to acq mode.
- *
- * \param[in]  channel  The channel to find the source for.
- * \return source for the channel.
- */
-static enum bl_acq_source device__get_channel_source(uint8_t channel)
+/* Exported function, documented in device.h */
+enum bl_acq_source device__get_channel_source(uint8_t channel)
```
Comment (tlsa) [inline host/bloodview/src/device.c:561]: Double underscore is for static functions!

```suggestion
enum bl_acq_source device_get_channel_source(uint8_t channel)
```

File: host/bloodview/src/device.h [hunk 4]
```diff
@@ -111,4 +111,11 @@ bool device_stop(
 const struct device_source_cap *device_get_source_cap(
 		enum bl_acq_source source);
 
+/**
+ * Get the appropriate source for a given channel, according to acq mode.
+ *
+ * \param[in]  channel  The channel to find the source for.
+ * \return source for the channel.
+ */
+enum bl_acq_source device__get_channel_source(uint8_t channel);
```
Comment (tlsa) [inline host/bloodview/src/device.h:120]:
```suggestion
enum bl_acq_source device_get_channel_source(uint8_t channel);
```

File: host/bloodview/src/data-cal.c [hunk 5]
```diff
@@ -290,15 +357,17 @@ void *data_cal_init(
 		return NULL;
 	}
 
-	ctx->channel = data_cal__init_channel_array(channels);
+	unsigned channel_count = __builtin_popcount(channel_mask);
```
Comment (tlsa) [inline host/bloodview/src/data-cal.c:360]: For the bloodview code, prefer to use `util_bit_count()` than something GCC-specific.

**Review by tlsa — APPROVED** [review]
Looks good!

File: host/bloodview/src/data-cal.c (hunk 1 above)
Comment (kplus) [reply host/bloodview/src/data-cal.c:284]: my local file seems to be OK, not sure if it's caused by editor setting or something...
will fix anyway

File: host/bloodview/src/device.c (hunk 3 above)
Comment (kplus) [reply host/bloodview/src/device.c:561]: +1

File: host/bloodview/src/device.h (hunk 4 above)
Comment (kplus) [reply host/bloodview/src/device.h:120]: +1

File: host/bloodview/src/data-cal.c (hunk 5 above)
Comment (kplus) [reply host/bloodview/src/data-cal.c:360]: +1

**Pull request merged**